### ISSUE TITLE
[RFC]: Classifier-Free Guidance

### ISSUE THREAD

**Issue by Vermeille** [vllm-project/vllm#5825]
### Motivation.

I am one of the authors of the paper Stay On Topic with Classifier-Free Guidance ( https://openreview.net/forum?id=RiM3cl9MdK&noteId=s1BXLL1YZD ) who has been nominated as ICML'24 Spotlight Paper. CFG is a sampling technique that allows LLMs to follow the prompt more closely at the cost of two forward passes per token as well as 2 kv caches. CFG brings non trivial improvements overall over standard benchmarks.

I would be extremely interested in having CFG implemented into vLLM. If possible, I would like to get a bit of guidance into the vLLM code base.

### Proposed Change.

CFG contrasts the next token logits between two different prompt (a "positive prompt" a, and a "negative prompt" or "unconditional" b)

Here is the pseudo algorithm

```
while we sample:
    logits_a = log_softmax(model(prompt_a))
    logits_b = log_softmax(model(prompt_b))
    logits = logits_b + cfg_scale * (logits_a - logits_b)
    next_token = sample_from(logits)
    prompt_a.append(next_token)
    prompt_b.append(next_token)
```

As you can see this needs two concurrent kv-caches for an efficient implementation. I tried looking for how Speculative Decoding was implemented but this was quite complex, more than CFG needs.

### Any Other Things.

I am willing to implement it myself given enough guidance as this looks like a non trivial thing to implement. I think something similar to / reusing bits of Speculative Decoding might be used but the code is non trivial.

**Comment by Vermeille** [comment]
Up

**Comment by cadedaniel** [comment]
Hi @Vermeille. Great work.

For implementation in vLLM, this can be done at a similar layer to Speculative Decoding:
```
LLMEngine
CFGWorker
< logic which calls the underlying worker twice, does logit math, samples >
Worker (2x)
```

The primary benefit of this design is that you can manage two block tables in the existing LLMEngine and scheduler without any modification. This is done in speculative decoding (with draft model) by splitting the KV cache space evenly into two equally-sized regions [1]. Then the same block table can work for both models. You can actually prototype this relatively straightforwardly; the only major missing piece is you will need to have one of the Workers not load weights (e.g. weight loading is shared with other worker).

Alternatively, you can use a single worker and modify block tables with a constant offset so that you have independent KV cache.

[1] https://github.com/vllm-project/vllm/blob/70c232f85a9e83421a4d9ca95e6384364271f2bc/vllm/spec_decode/spec_decode_worker.py#L255-L274
